Add ProtocolCache for 1Inch

MartinquaXD opened the pull request:
The `OneInchPriceEstimator` and `OneInchSolver` need to disable certain protocols (`PMM1`) because they can lead to wrong settlements.
To not duplicate the code responsible for limiting the protocols, I added `ProtocolCache` to `oneinch_api.rs` and made both structs us it.

This PR introduces caching to the `OneInchSolver` but it shouldn't be a big deal. 1Inch doesn't change the supported protocols on a regular basis, so having a small delay (at most 1 minute) until `OneInchSolver` is able to utilize new protocols can not lead to many suboptimal solutions.

### Test Plan
Unit tests for `ProtocolCache`

Commit:
Implement Cache for 1Inch protocols

Commit:
Implement tests for 1Inch ProtocolCache

Commit:
Use ProtocolCache in OneInchPriceEstimator

Commit:
Use ProtocolCache in OneInchSolver

vkgnosis commented on crates/shared/src/oneinch_api.rs:
```diff
@@ -384,6 +386,55 @@ where
     serde_json::from_str(&response?).context("1inch result parsing failed")
 }
 
+#[derive(Debug, Clone)]
+pub struct ProtocolCache(Arc<Mutex<TimedSizedCache<Vec<String>, Vec<String>>>>);
```
I think we should cache the raw list of protocols that 1nch returns instead of this construct where the entries in the cache depend on the disabled protocols. The underlying data is always the same so I imagine there would be a "get current protocols" function that either returns the cached value or updates the cache and returns the value.
That you might or might not implement with this cached trait. I'm saying might not because when there is only one value it might be so simple that the crate doesn't add anything.

nlordell replied:
One advantage of caching the protocol list instead of the enabled/disabled protocols is that it, in theory, allows multiple instances of `1Inch{PriceEstimator,Solver}` to share the same cache.

That being said, we don't really do that now so maybe its a bit of "overarchitecting" on my part.

> there is only one value it might be so simple that the crate doesn't add anything.

> But even if the general use case is only storing a single value, I would prefer using the crate over implementing a cache myself.

I don't have a strong opinion here. If we added `cached` as a dependency _just_ for this cache, I would tend to agree that we shouldn't use, but since its already included in our dependency tree - why not?

nlordell replied:
You could use `TimedCache<(), Vec<String>>` if you wanted to keep a single list right?

vkgnosis replied:
> I wanted to skip filtering out disabled protocols every time we need the allowed list

I like doing it that way a lot more. Filtering takes no time compared to making the network request and it has the advantage of simpler code because only one thing is cached.

nlordell commented on crates/shared/src/oneinch_api.rs:
```diff
@@ -384,6 +386,55 @@ where
+            return Ok(None);
+        }
+
+        if let Some(allowed_protocols) = self.0.lock().unwrap().cache_get(disabled_protocols) {
```
Interesting... This doesn't use `Borrow::Target` so you have to use a `&Vec<String>` instead of the more Rust-y `&[String]`.

Commit:
Cache all 1Inch protocols instead of filtered list

This makes the ProtocolCache more general purpose and cleans up the API
a bit.

nlordell commented on crates/shared/src/oneinch_api.rs:
```diff
@@ -384,6 +386,60 @@ where
 }
 
+#[derive(Debug, Clone)]
+pub struct ProtocolCache(Arc<Mutex<TimedSizedCache<(), Vec<String>>>>);
```
Can this just be `TimedCache` since the size can only ever be 1 so we don't care about it?

vkgnosis replied:
I think that is correct.

MartinquaXD replied:
They should behave the same in our case but their implementation is different. `SizedCache` and `TimedSizedCache` use quite a bit of `unsafe` in the authors homegrown implementation of the inner map storing the cached values.
`TimedCache` uses a simple `std::collections::HashMap` so let's go with that instead.

nlordell reviewed (approved):
Personally, I prefer it this way as well (caching a single value).

Also, generally speaking, very clean PR with some nice unit tests 🎉.

vkgnosis commented on crates/shared/src/oneinch_api.rs:
```diff
@@ -384,6 +386,60 @@ where
+pub struct ProtocolCache(Arc<Mutex<TimedSizedCache<(), Vec<String>>>>);
+
+impl ProtocolCache {
+    pub fn new(cache_validity_in_seconds: u64) -> Self {
```
Nit: Could be `Duration` and then use .as_secs() to pass into the cache constructor which sadly accepts only integer seconds instead of Duration.

vkgnosis commented on crates/shared/src/oneinch_api.rs:
```diff
@@ -384,6 +386,60 @@ where
+    }
+}
+
+impl Default for ProtocolCache {
+    fn default() -> Self {
+        Self::new(60)
```
Nit: We can make this a lot longer than 60 seconds. I don't expect them to add protocols often and even if they do it wouldn't be a problem to not use a new protocol right that moment. Could be an hour. On the other hand fetching the protocols isn't expensive either so not sure.

Commit:
Use TimedCache instead of TimedSizedCache

The "Sized" versions of the crates caches use quite a bit of unsafe
under the hood.
Since in our case TimedSizedCache<(), V> with size of 1 has an
equivalent behavior to TimedCache<(), V> with unbounded size let's go
with the safer version.
